Spark: Add location overlap validation for SnapshotTableAction

varun-lakhyani opened the pull request:
## Description
Implements location overlap validation for `SnapshotTableAction` to prevent destination table location from overlapping with source table location.

Resolves TODO comment in `SnapshotTableSparkAction.java:127` in spark v4.1.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- Added unit tests covering snapshot and source location overlap scenarios:
  - Exact same location
  - Source as a subdirectory of destination
  - Destination as a subdirectory of source
  - Completely non-overlapping locations
- All CI checks pass (31/31) 

## Checklist
- [x] Implementation follows existing code patterns and Iceberg contribution guidelines
- [x] Code compiles without errors
- [x] Test coverage added 
- [x] No breaking changes to public APIs

RussellSpitzer commented:
Definitely needs a test case, not sure why windows would make testing impossible since our locations are stored as strings

varun-lakhyani commented:
> Iceberg errors follow this format
> 
> "Cannot X because Y (possibly fix Z)"
> 
> So in this case
> 
> Cannot create a snapshot at location ... because it would overlap with source table ... Overlapping snapshot and source would ....

 @RussellSpitzer  
 I have addressed all your feedback - added tests and updated error formatting.
 All CI checks passing. Would appreciate a review when you are available, Happy to make any changes if required. 
 Thanks

varun-lakhyani commented:
@RussellSpitzer PTAL

varun-lakhyani commented:
PTAL @pvary

pvary commented on spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/actions/TestSnapshotTableAction.java:
```diff
@@ -65,4 +70,66 @@ public void testSnapshotWithParallelTasks() throws IOException {
+        .as(tableName)
+        .tableLocation(validDestLocation)
+        .execute();
+    assertThat(sql("SELECT * FROM %s", tableName)).hasSize(2);
```
Do we need this test here?
Should this be a different test case for the happy path?

varun-lakhyani commented on spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/actions/TestSnapshotTableAction.java:
```diff
@@ -65,4 +70,66 @@ public void testSnapshotWithParallelTasks() throws IOException {
+  @TestTemplate
+  public void testSnapshotWithOverlappingLocation() throws IOException {
+    String catalogType = catalogConfig.get(ICEBERG_CATALOG_TYPE);
+    assumeThat(catalogType).isNotEqualTo(ICEBERG_CATALOG_TYPE_HADOOP);
```
> Why do we don't test with hadoop?

Hadoop doesn't allow custom destination location via .tableLocation() when done this it fails even before the check while      `StagedSparkTable stagedTable = stageDestTable();`

Error Message:
`TestSnapshotTableAction > testSnapshotWithOverlappingLocation() > catalogName = testhadoop, implementation = org.apache.iceberg.spark.SparkCatalog, config = {type=hadoop, cache-enabled=false} FAILED
    java.lang.IllegalArgumentException: Cannot set a custom location for a path-based table. Expected file:/tmp/warehouse62334905503302693.tmp/default/table but got /tmp/junit-17415081344650343329/newJunit10343828823716601592`
        
because of HadoopCatalog withLocation() check

`public TableBuilder withLocation(String location) {
      Preconditions.checkArgument(
          location == null || location.equals(defaultLocation),
          "Cannot set a custom location for a path-based table. Expected "
              + defaultLocation
              + " but got "
              + location);
      return this;
    }`
    
So, when we can't give custom location and TestBase predefine the warehouse using which hadoop always derives non-overlapping location the test seems useless unless we change the warehouse.
For the same reason I skipped Hadoop catalog for the tests including .tableLocation() and didn't create another specially for it.
Do let me know If I should think otherwise

RussellSpitzer replied:
Best to just add a note here, "//Hadoop Catalogs do not Support Custom Table Locations"

varun-lakhyani commented on spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/actions/TestSnapshotTableAction.java:
```diff
@@ -18,12 +18,17 @@
 import java.nio.file.Files;
 import java.util.concurrent.Executors;
 import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.hadoop.fs.Path;
```
Yes Correct, It was just used for creating nested path locations.
Standard string operations are simpler and suitable than this. UPDATED in latest commit.
